tsl/src/fdw/data_node_scan_plan.c

```diff
@@ -674,6 +688,689 @@ push_down_group_bys(PlannerInfo *root, RelOptInfo *hyper_rel, Hyperspace *hs,
+
+	return join_partition;
+}
+
+/*
+ * Create a JoinPathExtraData data structure for a partiton. The new struct is based on the original
+ * JoinPathExtraData of the join and the AppendRelInfo of the partiton.
+ */
+static JoinPathExtraData *
+create_join_partition_extra(PlannerInfo *root, JoinPathExtraData *extra, AppendRelInfo *appinfo)
+{
+	JoinPathExtraData *partition_extra = palloc(sizeof(JoinPathExtraData));
+	partition_extra->inner_unique = extra->inner_unique;
+	partition_extra->sjinfo = extra->sjinfo;
+	partition_extra->semifactors = extra->semifactors;
+	partition_extra->param_source_rels = extra->param_source_rels;
+	partition_extra->mergeclause_list =
+		castNode(List, adjust_appendrel_attrs(root, (Node *) extra->mergeclause_list, 1, &appinfo));
+	partition_extra->restrictlist =
+		castNode(List, adjust_appendrel_attrs(root, (Node *) extra->restrictlist, 1, &appinfo));
+
+	return partition_extra;
+}
+
+/*
+ * Generate the paths for a pushed down join. Each data node will be considered as a partition
+ * of the join. The join can be pushed down if:
+ *
+ * (1) The setting "ts_guc_enable_per_data_node_queries" is enabled
+ * (2) The outer relation is a distributed hypertable
+ * (3) The inner relation is marked as a reference table
+ * (4) The join is a left join or an inner join
+ *
+ * The join will be performed between the multiple DataNodeRels (see function
+ * build_data_node_part_rels) and the original innerrel of the join (the reftable).
+ */
+void
+data_node_generate_pushdown_join_paths(PlannerInfo *root, RelOptInfo *joinrel, RelOptInfo *outerrel,
+									   RelOptInfo *innerrel, JoinType jointype,
+									   JoinPathExtraData *extra)
+{
+	TsFdwRelInfo *fpinfo;
+	Path *joinpath;
+	double rows = 0;
+	int width = 0;
+	Cost startup_cost = 0;
+	Cost total_cost = 0;
+	Path *epq_path = NULL;
+	RelOptInfo **hyper_table_rels;
+	RelOptInfo **join_partition_rels;
+	int nhyper_table_rels;
+	List *join_part_rels_list = NIL;
+#if PG15_GE
+	Bitmapset *data_node_live_rels = NULL;
+#endif
+
+	/*
+	 * Skip check if the join result has been considered already.
+	 */
+	if (joinrel->fdw_private)
+		return;
+
+	/*
+	 * This code does not work for joins with lateral references, since those
+	 * must have parameterized paths, which we don't generate yet.
+	 */
+	if (!bms_is_empty(joinrel->lateral_relids))
+		return;
+
+	/* The outerrel has to be distributed. This condition should be always hold
+	 * because otherwise we should not start the planning for distributed tables. However
+	 * we can check this condition here and fail more nicely by doing nothing than hitting
+	 * a failed assert. */
+	TimescaleDBPrivate *outerrel_private = outerrel->fdw_private;
+	if (outerrel_private == NULL || outerrel_private->fdw_relation_info == NULL)
+		return;
+
+	/* Ensure pushdown planning is only performed on the AN. On the DNs the outererl is not
+	 * partitioned. */
+	if (outerrel->nparts <= 0)
```

Comment:
I guess this code isn't even invoked on data nodes, so we can probably assert this.

Reply:
I have seen join replanning on the DNs. So, I added this code to ensure that only the AN performs the planning.

Reply:
> I have seen join replanning on the DNs.

This would be a serious bug, the data nodes don't use data node scans. Let's make this an Assert or Ensure.

Reply:
So far, the function was called by the generic planner hook every time a join was part of the query and `fdw_private` was not empty. I made the generic planner hook (`timescaledb_set_join_pathlist_hook`) more selective when this FDW function is called. With this change, I could convert this statement to an Assert.